File: backend/src/controllers/authController.js
const bcrypt = require('bcryptjs');
const jwt = require('jsonwebtoken');
const { validationResult } = require('express-validator');
const User = require('../models/User');
const { query } = require('../config/db');

const generateRenewalNotifications = async (userId, isAdmin = false) => {
  try {
    const params = [];
    let agentFilter = '';

    if (!isAdmin) {
      agentFilter = 'AND p.agent_id = ?';
      params.push(userId);
    }

    const policies = await query(
      `SELECT p.policy_id, p.customer_name, p.policy_number, p.end_date, p.agent_id,
              p.renewal_notified_30, p.renewal_notified_60, p.renewal_notified_90,
              DATEDIFF(p.end_date, CURDATE()) AS days_to_expiry
       FROM policies p
       WHERE p.end_date BETWEEN CURDATE() AND DATE_ADD(CURDATE(), INTERVAL 90 DAY)
       ${agentFilter}`,
      params
    );

    for (const policy of policies) {
      const { days_to_expiry, policy_id, customer_name, policy_number, agent_id } = policy;

      if (days_to_expiry <= 30 && !policy.renewal_notified_30) {
        await query(
          "INSERT INTO notifications (user_id, message, type, status) VALUES (?, ?, 'renewal', 'unread')",
          [agent_id, `Policy ${policy_number} for ${customer_name} expires in ${days_to_expiry} day(s). Please initiate renewal.`]
        );
        await query('UPDATE policies SET renewal_notified_30 = 1 WHERE policy_id = ?', [policy_id]);
      } else if (days_to_expiry <= 60 && !policy.renewal_notified_60) {
        await query(
          "INSERT INTO notifications (user_id, message, type, status) VALUES (?, ?, 'renewal', 'unread')",
          [agent_id, `Policy ${policy_number} for ${customer_name} expires in ${days_to_expiry} day(s). Please plan renewal.`]
        );
        await query('UPDATE policies SET renewal_notified_60 = 1 WHERE policy_id = ?', [policy_id]);
      } else if (days_to_expiry <= 90 && !policy.renewal_notified_90) {
        await query(
          "INSERT INTO notifications (user_id, message, type, status) VALUES (?, ?, 'renewal', 'unread')",
          [agent_id, `Policy ${policy_number} for ${customer_name} expires in ${days_to_expiry} day(s). Renewal due soon.`]
        );
        await query('UPDATE policies SET renewal_notified_90 = 1 WHERE policy_id = ?', [policy_id]);
      }
    }
  } catch (err) {
    console.error('generateRenewalNotifications error:', err.message);
  }
};

const login = async (req, res, next) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ success: false, message: 'Validation failed', errors: errors.array() });
    }

    const { email, password } = req.body;

    const user = await User.findByEmail(email);
    if (!user) {
      return res.status(401).json({ success: false, message: 'Invalid email or password' });
    }

    if (!user.is_active) {
      return res.status(401).json({ success: false, message: 'Account is inactive' });
    }

    const passwordMatch = await bcrypt.compare(password, user.password);
    if (!passwordMatch) {
      return res.status(401).json({ success: false, message: 'Invalid email or password' });
    }

    const payload = {
      user_id: user.user_id,
      email: user.email,
      role_id: user.role_id,
      role_name: user.role_name,
      branch_id: user.branch_id
    };

    const token = jwt.sign(payload, process.env.JWT_SECRET, {
      expiresIn: process.env.JWT_EXPIRES_IN || '24h'
    });

    // Generate renewal notifications in background
    const isAdmin = user.role_name === 'Admin';
    generateRenewalNotifications(user.user_id, isAdmin).catch(() => {});

    return res.status(200).json({
      success: true,
      token,
      user: {
        user_id: user.user_id,
        name: user.name,
        email: user.email,
        role_name: user.role_name,
        role_id: user.role_id,
        branch_id: user.branch_id,
        branch_name: user.branch_name
      }
    });
  } catch (err) {
    next(err);
  }
};

const getMe = async (req, res, next) => {
  try {
    const user = await User.findById(req.user.user_id);
    if (!user) {
      return res.status(404).json({ success: false, message: 'User not found' });
    }
    const { password: _pw, ...safeUser } = user;
    return res.status(200).json({ success: true, data: safeUser });
  } catch (err) {
    next(err);
  }
};

const logout = async (req, res) => {
  return res.status(200).json({ success: true, message: 'Logged out successfully' });
};

const changePassword = async (req, res, next) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ success: false, message: 'Validation failed', errors: errors.array() });
    }

    const { currentPassword, newPassword } = req.body;

    const rows = await query('SELECT password FROM users WHERE user_id = ?', [req.user.user_id]);
    if (!rows.length) {
      return res.status(404).json({ success: false, message: 'User not found' });
    }

    const passwordMatch = await bcrypt.compare(currentPassword, rows[0].password);
    if (!passwordMatch) {
      return res.status(400).json({ success: false, message: 'Current password is incorrect' });
    }

    const hashed = await bcrypt.hash(newPassword, 10);
    await query('UPDATE users SET password = ? WHERE user_id = ?', [hashed, req.user.user_id]);

    return res.status(200).json({ success: true, message: 'Password changed successfully' });
  } catch (err) {
    next(err);
  }
};

module.exports = { login, getMe, logout, changePassword, generateRenewalNotifications };

File: dotnet-api/Controllers/AuthController.cs
using ActivityTrackerAPI.Helpers;
using ActivityTrackerAPI.Models.DTOs;
using ActivityTrackerAPI.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace ActivityTrackerAPI.Controllers;

/// <summary>Authentication endpoints</summary>
[ApiController]
[Route("api/auth")]
[Produces("application/json")]
public class AuthController : ControllerBase
{
    private readonly IUserService _userService;
    private readonly IJwtService _jwtService;
    private readonly INotificationService _notificationService;
    private readonly IPolicyService _policyService;
    private readonly IServiceProvider _serviceProvider;

    public AuthController(
        IUserService userService,
        IJwtService jwtService,
        INotificationService notificationService,
        IPolicyService policyService,
        IServiceProvider serviceProvider)
    {
        _userService = userService;
        _jwtService = jwtService;
        _notificationService = notificationService;
        _policyService = policyService;
        _serviceProvider = serviceProvider;
    }

    /// <summary>Login with email and password</summary>
    [HttpPost("login")]
    [AllowAnonymous]
    [ProducesResponseType(typeof(LoginResponse), 200)]
    [ProducesResponseType(401)]
    public async Task<IActionResult> Login([FromBody] LoginRequest request)
    {
        if (!ModelState.IsValid)
            return BadRequest(new { success = false, message = "Validation failed", errors = ModelState });

        var user = await _userService.GetByEmailAsync(request.Email);
        if (user == null)
            return Unauthorized(new { success = false, message = "Invalid email or password" });

        if (!user.IsActive)
            return Unauthorized(new { success = false, message = "Account is inactive" });

        if (!BCrypt.Net.BCrypt.Verify(request.Password, user.Password))
            return Unauthorized(new { success = false, message = "Invalid email or password" });

        var token = _jwtService.GenerateToken(user);

        // Fire-and-forget renewal notifications (scoped services via IServiceProvider)
        var isAdmin = user.RoleName == "Admin";
        var capturedUserId = user.UserId;
        var sp = _serviceProvider;
        _ = Task.Run(async () =>
        {
            try
            {
                using var scope = sp.CreateScope();
                var policyService = scope.ServiceProvider.GetRequiredService<IPolicyService>();
                var notifService = scope.ServiceProvider.GetRequiredService<INotificationService>();
                await GenerateRenewalNotificationsAsync(policyService, notifService, capturedUserId, isAdmin);
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine($"generateRenewalNotifications error: {ex.Message}");
            }
        });

        return Ok(new
        {
            success = true,
            token,
            user = new
            {
                user_id = user.UserId,
                name = user.Name,
                email = user.Email,
                role_name = user.RoleName,
                role_id = user.RoleId,
                branch_id = user.BranchId,
                branch_name = user.BranchName
            }
        });
    }

    /// <summary>Get current authenticated user profile</summary>
    [HttpGet("me")]
    [Authorize]
    [ProducesResponseType(200)]
    [ProducesResponseType(401)]
    public async Task<IActionResult> GetMe()
    {
        var userId = User.GetUserId();
        var user = await _userService.GetByIdAsync(userId);
        if (user == null)
            return NotFound(new { success = false, message = "User not found" });

        return Ok(new
        {
            success = true,
            data = new
            {
                user.UserId,
                user.Name,
                user.Email,
                user.RoleId,
                user.RoleName,
                user.BranchId,
                user.BranchName,
                user.IsActive,
                user.CreatedAt,
                user.Permissions
            }
        });
    }

    /// <summary>Logout (client-side token invalidation)</summary>
    [HttpPost("logout")]
    [Authorize]
    [ProducesResponseType(200)]
    public IActionResult Logout()
    {
        return Ok(new { success = true, message = "Logged out successfully" });
    }

    /// <summary>Change current user's password</summary>
    [HttpPost("change-password")]
    [Authorize]
    [ProducesResponseType(200)]
    [ProducesResponseType(400)]
    public async Task<IActionResult> ChangePassword([FromBody] ChangePasswordRequest request)
    {
        if (!ModelState.IsValid)
            return BadRequest(new { success = false, message = "Validation failed", errors = ModelState });

        var userId = User.GetUserId();
        var currentPassword = await _userService.GetPasswordAsync(userId);
        if (currentPassword == null)
            return NotFound(new { success = false, message = "User not found" });

        if (!BCrypt.Net.BCrypt.Verify(request.CurrentPassword, currentPassword))
            return BadRequest(new { success = false, message = "Current password is incorrect" });

        var hashed = BCrypt.Net.BCrypt.HashPassword(request.NewPassword);
        await _userService.UpdatePasswordAsync(userId, hashed);

        return Ok(new { success = true, message = "Password changed successfully" });
    }

    private static async Task GenerateRenewalNotificationsAsync(
        IPolicyService policyService,
        INotificationService notifService,
        uint userId,
        bool isAdmin)
    {
        var policies = await policyService.GetRenewalPoliciesForNotificationsAsync(
            isAdmin ? null : userId);

        foreach (var policy in policies)
        {
            if (policy.DaysToExpiry <= 30 && !policy.RenewalNotified30)
            {
                await notifService.CreateAsync(policy.AgentId,
                    $"Policy {policy.PolicyNumber} for {policy.CustomerName} expires in {policy.DaysToExpiry} day(s). Please initiate renewal.",
                    "renewal");
                await policyService.MarkRenewalNotifiedAsync(policy.PolicyId, 30);
            }
            else if (policy.DaysToExpiry <= 60 && !policy.RenewalNotified60)
            {
                await notifService.CreateAsync(policy.AgentId,
                    $"Policy {policy.PolicyNumber} for {policy.CustomerName} expires in {policy.DaysToExpiry} day(s). Please plan renewal.",
                    "renewal");
                await policyService.MarkRenewalNotifiedAsync(policy.PolicyId, 60);
            }
            else if (policy.DaysToExpiry <= 90 && !policy.RenewalNotified90)
            {
                await notifService.CreateAsync(policy.AgentId,
                    $"Policy {policy.PolicyNumber} for {policy.CustomerName} expires in {policy.DaysToExpiry} day(s). Renewal due soon.",
                    "renewal");
                await policyService.MarkRenewalNotifiedAsync(policy.PolicyId, 90);
            }
        }
    }
}

